projects/sandbox/train/train/validation.py

```diff
@@ -245,299 +50,220 @@ class Recorder:
-
-        # sample waveforms and rescale snrs below threshold
-        waveforms, _ = injector.sample(-1)
-        df = 1 / (waveforms.shape[-1] / sample_rate)
-        psds = []
-        for back in background:
-            psd = normalize_psd(back, df, sample_rate)
-            psds.append(psd)
-        psds = torch.tensor(np.stack(psds), dtype=torch.float64)
-
-        snrs = compute_network_snr(waveforms, psds, sample_rate, highpass)
-        mask = snrs < snr_thresh
-        logging.info(
-            f"Rescaling {mask.sum()} out of {len(snrs)} "
-            "waveforms below snr threshold"
-        )
-        snrs[mask] = snr_thresh
-        waveforms = reweight_snrs(waveforms, snrs, psds, sample_rate, highpass)
-
-        # create a dataset of pure background
-        background = make_background(background, kernel_size, stride_size)
-        self.background_loader = self.make_loader(background, batch_size)
-
-        # now repliate that dataset but with glitches inserted
-        # into either or both interferometer channels
-        glitch_background = make_glitches(glitches, background, glitch_frac)
-        self.glitch_loader = self.make_loader(glitch_background, batch_size)
-
-        # create a tensor of background with waveforms injected.
-        signal_background = repeat(background, len(waveforms))
-
-        start = waveforms.shape[-1] // 2 - kernel_size // 2
-        stop = start + kernel_size
-        signal_background += waveforms[:, :, start:stop]
-        self.signal_loader = self.make_loader(signal_background, batch_size)
-
-    def make_loader(self, X: torch.Tensor, batch_size: int):
-        dataset = torch.utils.data.TensorDataset(X)
-        return torch.utils.data.DataLoader(
-            dataset,
-            pin_memory=True,
-            batch_size=batch_size,
-            pin_memory_device=self.device,
-        )
+        preds = torch.cat(preds)
+        if inj_preds:
+            inj_preds = torch.cat(inj_preds)
+        else:
+            inj_preds = None
 
-    def get_predictions(
-        self, loader: Iterable[Tuple[torch.Tensor]], model: torch.nn.Module
-    ) -> torch.Tensor:
-        preds = []
-        for (X,) in loader:
-            X = X.to(self.device)
-            y_hat = model(X)[:, 0]
-            preds.append(y_hat)
-        return torch.cat(preds)
+        preds = self.postprocess(preds)
+        return preds, inj_preds
 
-    @torch.no_grad()
     def __call__(self, model: torch.nn.Module, train_loss: float) -> bool:
-        background_preds = self.get_predictions(self.background_loader, model)
-        glitch_preds = self.get_predictions(self.glitch_loader, model)
-        signal_preds = self.get_predictions(self.signal_loader, model)
-
-        return self.recorder(
-            model, train_loss, background_preds, glitch_preds, signal_preds
-        )
+        model.eval()
+        self._injection_idx = 0
+
+        # do the 0-shift inference manually since
```

Comment:
Probably should just skip the 0-shift right. We are collecting background so no need in obfuscating that. Also, in general, may be some signals hidden there.

Reply:
Yeah true great call